optaweb-employee-rostering-frontend/src/ui/components/WeekPicker.tsx

```diff
@@ -40,28 +35,71 @@ function getLastDayInWeek(dateInWeek: Date): Date {
   return moment(dateInWeek).endOf('week').toDate();
 }
 
-export default class WeekPicker extends React.Component<WeekPickerProps, WeekPickerState> {
-  constructor(props: WeekPickerProps) {
-    super(props);
-    this.state = { isOpen: false };
-  }
-
+export default class WeekPicker extends React.Component<WeekPickerProps> {
   goToCurrentWeek() {
     this.goToWeekContaining(new Date());
   }
 
   goToWeekContaining(date: Date) {
     if (this.props.minDate && getFirstDayInWeek(date) < getFirstDayInWeek(this.props.minDate)) {
       this.goToWeekContaining(this.props.minDate);
-    } else if (this.props.maxDate && getLastDayInWeek(date) > getLastDayInWeek(this.props.maxDate)) {
+      return;
+    } if (this.props.maxDate && getLastDayInWeek(date) > getLastDayInWeek(this.props.maxDate)) {
       this.goToWeekContaining(this.props.maxDate);
+      return;
     }
     this.props.onChange(getFirstDayInWeek(date), getLastDayInWeek(date));
-    this.setState({ isOpen: false });
   }
 
   render() {
-    const locale = moment.locale();
+    type DatePickerCustomInputProps = { value?: Date; onClick?: () => void };
+
+    // (the stateless functional component is defined in a class component, where
+    //  this = the class this, so it okay to use "this" here)
+    /* eslint-disable react/no-this-in-sfc */
+    const DateRangeInputElement = React.forwardRef<HTMLDivElement, DatePickerCustomInputProps>(
```

Comment:
Could you please move this component to a dedicated file? The render method is already pretty long so moving this sub-component outside will make the WeekPicker easier to understand. Also, the move will help to get rid of the references to `this` and replace them by props. A function component output should be fully determined by the input props (and perhaps its state).

More about this at https://reactjs.org/docs/components-and-props.html.

Reply:
I disagree with moving it to another file; the custom input component is customized for use with the week picker component and wouldn't make sense using it in another component/page. I can extract the component though (won't be in the class render method, but be in the same file).

Reply:
That also works.

Generally, I don't think that reusing a component more than once is a necessary requirement for moving it to a separate file. One component per file helps to keep the file size limited making it easier to read and focus on a specific problem the component solves.